tests/unit/notification/QueueProcessor.spec.js

```diff
@@ -0,0 +1,151 @@
+    // test was written Bucket Notification tests are not yet running
+    // in Integration (refs: BB-419, INTGR-895)
+    it('should process 10K notifications with 1K concurrency to send to external destination',
+    function test10KNotif(done) {
+        this.timeout(5000);
+
+        const origLogger = qp.logger;
+        qp.logger = {
+            debug: () => {},
+            info: () => {},
+        };
+        // fake how Kafka Producer delivers pending delivery reports at a regular interval
+        let pendingDeliveryReports = [];
+        const flushDeliveryReportsInterval = setInterval(() => {
+            const _pendingDeliveryReports = pendingDeliveryReports;
+            pendingDeliveryReports = [];
+            for (const deliveryReportCb of _pendingDeliveryReports) {
+                deliveryReportCb();
+            }
+        }, 200);
+        const send = sinon.stub().callsFake((messages, cb) => {
+            // push the delivery report to be delivered asynchronously later
+            pendingDeliveryReports.push(cb);
+        });
+        qp._destination._notificationProducer = {
+            send,
+        };
+        async.timesLimit(10000, 1000, (i, iDone) => qp.processKafkaEntry({
+            value: JSON.stringify({
+                bucket: 'mybucket',
+                key: `key-${i}`,
+                eventType: 's3:ObjectCreated:Put',
+                value: '{}',
+            }),
+        }, iDone), err => {
+            assert.ifError(err);
+            assert.strictEqual(pendingDeliveryReports.length, 0);
+            qp.logger = origLogger;
+            clearInterval(flushDeliveryReportsInterval);
+            done();
+        });
+    });
+
+    it('should send notification to external destination with delivery error', done => {
+        const send = sinon.stub().callsFake((messages, cb) => cb(new Error('delivery error')));
+        qp._destination._notificationProducer = {
+            send,
+        };
+        qp.processKafkaEntry({
+            value: JSON.stringify({
+                bucket: 'mybucket',
+                key: 'key',
+                eventType: 's3:ObjectCreated:Put',
+                value: '{}',
+            }),
+        }, err => {
+            assert.ifError(err);
+
+            assert(send.calledOnce);
+            setTimeout(done, 100);
+        });
+    });
+
+    it('should not send an entry without "eventType" attribute', done => {
+        const send = sinon.spy();
+        qp._destination._notificationProducer = {
+            send,
+        };
+        qp.processKafkaEntry({
+            value: JSON.stringify({
+                bucket: 'mybucket',
+                key: 'key',
+                // no "eventType"
+                value: '{}',
+            }),
+        }, err => {
+            assert.ifError(err);
+            // should not have been sent
+            assert(!send.calledOnce);
+            done();
+        });
+    });
+});
```

Comment:
is it possible that JSON parsing might fail in qp.processKafkaEntry? if so, lets add a test case

Reply:
It actually throws an exception if the input is not valid JSON.

It's not directly related to the scope of this PR, but it can be a good idea to return a proper error instead and add a test case for it indeed.

Reply:
I fixed JSON parsing exception and added a test